### PR TITLE
Automatically set TSC frequency and label nodes with the correct scheduling info

### PR DESCRIPTION
**Pull request by rmohr**
**What this PR does / why we need it**:

This is the second part of #5654. #5654 exposed the necessary TSC info on the nodes which are now used in this PR to allow scheduling VMIs with the right frequency and allow migrationg the VMIs with a frequency set.

VMIs with the invtsc cpuflag can only be migrated if they have a TSC frequency set. As a consequence this PR does the following:

***Labeling nodes***

In #5654 TSC frequency and scalability was exposed. However this information can not be used for scheduling. To support scheduling, out of this information schedulable TSC flags are calculated in this PR and set by virt-controller on the nodes.
For that, the lowest TSC frequency in the cluster is evaluated every 30 seconds. Then the frequency is added to every node with scalable TSC in the form `scheduling.node.kubevirt.io/tsc-frequency-<frequency>`. If the lowest frequency in the cluster is `123`, then the label has the form `scheduling.node.kubevirt.io/tsc-frequency-123`. Since the calculation re-done every 30 seconds, the controller follows node-topology changes with a maximum delay of 30 seconds.

In case that a VMI got scheduled with a frequency of `123`, but the minimum available frequency changes to `100` since another node got added which has this frequency, `virt-controller` keeps track of this already assigned frequency and nodes which suppoort it will get the new  `scheduling.node.kubevirt.io/tsc-frequency-100` frequency but also keep `scheduling.node.kubevirt.io/tsc-frequency-123` until the last VMI with that frequency is shut down.

***Setting the frequency on the VMI***

That a VMI gets a tsc frequency set, the following preconditions must be set:

 * kubevirt runs on amd64 (to my knowledge the only architecture where this is fully tested and supposed to work).
 * the VMI has an eviction strategy of `LiveMigrate`. Otherwise the VMI will be marked as not migratable.
 * The `invtsc` cpuflag must be requested.

If these preconditions are met, virt-controller will first populate the topology hints including the TSC frequency on the VMI status. In the next step it sets the TSC frequency scheduling label on the launcher pod. Finally virt-handler picks up that frequency from the VMI status and instructs libvirt to set the appropriate frequency on the domain XML.

***Advance configuration of the node topology updater***

In some cases it may be necessary to set a minimum TSC cluster frequency due to misreporting nodes statically. This can be done via the `spec.configuration.developerConfiguration.minimumClusterTSCFrequency` setting. Once set, on the next topology update round, all nodes which can support this frequency will be labelled with it and it will be treated as the new minimum frequency for the whole cluster.

In case that the topology updater misbehaves, `virt-controller` has a new commandline flag called `--node-topology-update-perio`. It can be set via our generic patching mechanism for our deployments.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
VMIs which choose evictionStrategy `LifeMigrate` and request the `invtsc` cpuflag are now live-migrateable
```

### CODE REVIEW

File: pkg/virt-controller/watch/topology/nodetopologyupdater.go
Comment (davidvossel) [inline pkg/virt-controller/watch/topology/nodetopologyupdater.go:95]: this might be mostly theoretical, but i think there is a race condition here with this label removal.

if a vmi doesn't have topology hints assigned it could grab something from the node informer cache for scheduling and set it on the VMI while this node topology updater is removing it.  There's a small gap due to the node cache lagging behind that could result in the VMI getting a frequency that that doesn't have a matching label on a node anymore.

it might not be practical to close this gap.  It's just something i noticed.

**Review by davidvossel — COMMENTED** [review]
did a quick pass. left some comments

File: pkg/virt-controller/watch/topology/hinter.go
Comment (rmohr) [reply pkg/virt-controller/watch/topology/hinter.go:29]: The main reason was that I am not sure if I can assume that all nodes have the same TSC frequency or have all scalable TSC. The `EvictionStrategy` was the closest thinkg to `this VMI must be migratable`. Therefore I chose it, to not limit VMs to a possibly small subset of nodes while they could run on all of them if the VMI does not have to be migratable.

File: pkg/virt-controller/watch/topology/nodetopologyupdater.go
Comment (rmohr) [reply pkg/virt-controller/watch/topology/nodetopologyupdater.go:95]: That should not be an issue. The worst thing that would happen is that the VMI can't be scheduled for 30 seconds, because on the reevaluation the topologyupdater would see the correct values on the VMI and would re-add the labels where possible.

File: pkg/virt-controller/watch/application.go
Comment (rmohr) [reply pkg/virt-controller/watch/application.go:426]: I was lazy and added it for unit test convenience. Let me re-evaluate what to do instead.

File: pkg/virt-controller/watch/application.go
Comment (rmohr) [reply pkg/virt-controller/watch/application.go:425]: I can move it to the run function. Will do that.

File: pkg/virt-controller/watch/topology/hinter.go
Comment (davidvossel) [reply pkg/virt-controller/watch/topology/hinter.go:34]: yeah, i get what you're thinking. I'm not sure how I feel about it though now that we have a controller (workload update controller) that can automate live migrations for workloads that don't necessarily have EvictionStrategy: LiveMigrate set. 

i'd rather we were inefficient in the TSC calculations than for us to create a coupling between evictionStrategy and TSC that might not hold up long term.

File: pkg/virt-controller/watch/topology/nodetopologyupdater.go
Comment (davidvossel) [reply pkg/virt-controller/watch/topology/nodetopologyupdater.go:107]: +1

File: pkg/virt-controller/watch/vmi.go
Comment (davidvossel) [inline pkg/virt-controller/watch/vmi.go:751]: this event is probably getting fired twice on every reconcile loop for a VMI when topology hints can't be found. 

maybe we could keep this

File: pkg/virt-controller/watch/vmi.go
Comment (davidvossel) [inline pkg/virt-controller/watch/vmi.go:749]: could we make a function in topologyHinter called `topologyHinter.RequiresTopologyHints(vmi)` which just tells us if topology hints are required for the VMI or not.

Then this if check could be

```
if vmi.Status.TopologyHints == nil && c.topologyHinter.TopologyHintsRequiredForVMI(vmi) {
    log.Log.V(3).Object(vmi).Infof("Delaying pod creation until topology hints are set")
    // no hints in the status yet, but apparently the VMI requires some, let's retry when they are there
    return nil
}
```

One reason I suggest that is that this logic is competing with UpdateStatus a bit. In the event that the topology hints aren't found, I think we're firing this ```FailedGatherhingClusterTopologyHints``` event twice for each reconcile loop.

**Review by davidvossel — COMMENTED** [review]
I made one suggestion that might simplify vmi.go a bit.

Other than that. I think the only point of contention left to discuss on this before merging is the `EvictionStrategy: LiveMigrate` behavior. 

I still don't see the value in keying of `EvictionStrategy: LiveMigrate` in order to enable TSC frequency scheduling. 

For TSC scheduling, regardless of live migration, we're still only talking about the subset of VMIs that have that CPU feature enabled. I'd be in favor of always performing TSC scheduling now, and optimizing in the future if we find we need to for some reason. 

I don't want to set the precedence that `EvictionStrategy: LiveMigrate` means anything beyond the behavior when the Eviction API is called on a VMI pod.

File: pkg/virt-controller/watch/vmi.go
Comment (rmohr) [reply pkg/virt-controller/watch/vmi.go]: :+1: Fixed.

**Comment by rmohr** [comment]
> I don't want to set the precedence that `EvictionStrategy: LiveMigrate` means anything beyond the behavior when the Eviction API is called on a VMI pod.

:+1 Removed.

File: pkg/virt-controller/watch/vmi.go
Comment (davidvossel) [inline pkg/virt-controller/watch/vmi.go:753]: do you still need this logic here that actually generates the hints for the VMI now that the code checks if the toplogy hints are nil && required? 

I think this is still causing multiple Events to get fired every reconcile loop. One in sync and another in the status update.

File: pkg/virt-handler/vm.go
Comment (davidvossel) [inline pkg/virt-handler/vm.go:1298]: is this possible now that the VMI won't get scheduled unless the hints exist?

Comment (rmohr) [reply pkg/virt-handler/vm.go:1298]: I am not sure how practically relevant it is, but if someone has a VM running nowadays it would set the right condition. However it would fail migrations nowadays too already. Will remove it.

File: pkg/virt-controller/watch/vmi.go
Comment (rmohr) [reply pkg/virt-controller/watch/vmi.go:753]: Yes the purpose was to ensure that if for some reason a pod can't be created for some time that the hints can still be updated. Will remove it. It is probably over-optimized.

**Comment by rmohr** [comment]
All points addressed.

**Review by davidvossel — APPROVED** [review]
/lgtm
/approve

**Pull request merged**
